src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1277,287 +804,149 @@ where
-                    // TODO(guswynn&aljoscha): Calculate batch_upper once. :D
-                    let reclock_result =
-                        timestamper.reclock_compat(&mut untimestamped_batch.messages);
-                    let reclocked = match reclock_result {
-                        Ok(reclocked) => reclocked,
-                        Err(ReclockError::BeyondUpper(_) | ReclockError::Uninitialized) => {
-                            trace!(
-                                "reclock({id}) {worker_id}/{worker_count}: \
-                                cannot yet reclock batch with \
-                                batch_upper: {:?} \
-                                batch_lower: {:?} \
-                                reclock.source_frontier: {:?}",
-                                untimestamped_batch.batch_upper,
-                                untimestamped_batch.batch_lower,
-                                timestamper.source_upper()
-                            );
-                            // We keep batches in the order they arrive from the
-                            // source. And we assume that the source frontier never
-                            // regressses. So we can break now.
-                            break;
-                        }
-                        Err(ReclockError::NotBeyondSince(ts)) => {
-                            panic!("failed to reclock time {:?}", ts)
-                        }
-                    };
+                    // If the remap frontier advanced it's time to carve out a batch that includes
+                    // all updates not beyond the upper
+                    Event::Progress(remap_upper) => {
+                        let remap_trace_batch = ReclockBatch {
+                            updates: remap_updates_stash
+                                .drain_filter_swapping(|(_, ts, _)| !remap_upper.less_equal(ts))
+                                .collect(),
+                            upper: remap_upper.to_owned(),
+                        };
+                        trace!("timely-{worker_id} reclock({id}) received remap batch: updates={:?} upper={}", &remap_trace_batch.updates, remap_upper.pretty());
+                        timestamper.push_trace_batch(remap_trace_batch);
+                        work_to_do.notify_one();
+                    }
+                },
+                Some(event) = source_rx.recv() => match event {
+                    EventCore::Progress(changes) => {
+                        source_upper.update_iter(changes);
+                        trace!("timely-{worker_id} reclock({id}) received source progress: source_upper={}", source_upper.pretty());
+                        work_to_do.notify_one();
+                    }
+                    EventCore::Messages(_time, batch) => {
+                        untimestamped_batches.push(batch)
+                    }
+                },
+                _ = work_to_do.notified(), if timestamper.initialized() => {
```

Comment:
why do we do this in a separate branch? I think it would cleaner to do it in the `EventCore::Progress` branch, and just move the `if timestamper.intialized()` statement into there? The `ready_upper` would be effectively contain ALL messages, once we are initialized

Reply:
Because we need to do work when either of the two branches above have their frontier advanced so I can't put this code in either of them